### DISCUSSION TITLE
Using minidsp-rs in Docker

### DISCUSSION THREAD

**Discussion by tmowbrey** [mrene/minidsp-rs#104] [Q&A]
Hello,

I have spun up a debian container in docker, installed minidsp-rs and connect to minidsp but do not get the behavior I'm expecting. I'm curious is there are any thoughts on what is happening.

Additional background:

1. I am using MiniDSP 2x4 HD + WI-DG
2. I have successfully installed the 0.0.7 deb package
3. I can successfully run the Minidsp 2x4 HD Plugin software on my laptop (this is not the docker container as it doesn't have a UI) and connect and manipulate the Minidsp's configuration
4. I can successfully ping the WI-DG from the docker container. This goes through the following hops: 10.0.3.2 (docker container) --> 10.0.3.1 (docker gateway) --> 192.168.1.20 (NAS) --> 192.168.1.183 (WI-DG)
5. I have ports 5333 and 3999 mapped between the NAS host and the docker container
6. When I run "minidsp --tcp=192.168.1.183 -v" I get the following: 
```
Sent: ReadHardwareId
Recv: HardwareId { payload: [01, 0c, 0a] }
Sent: ReadMemory { addr: ffa1, size: 01 }
```

I really appreciate any thoughts or support anyone could provide with this. Thanks!

**Comment by mrene** [comment]
Hmm network-wise it looks fine since it's getting the hardware id back. 

Is the plugin app open at the same time? I don't have a WI-DG so I have no idea whether they tolerate multiple clients correctly, or if responses to commands end up at the right connection.

If not, I've seen devices get in a broken state sometimes, does restarting your 2x4hd solves the issue?

**Reply by tmowbrey** [reply]
So the process I got through is to check my connection with the plugin app and pushing my saved config to the Minidsp. This happens successfully so then I close the plugin app and then run minidsp-rs after that and get the output above.

As an aside, I've just finished installing the windows app on my laptop and get the same output (removing the container networking out of the equation for easier troubleshooting).

I'll try rebooting in about 30 minutes and get back to you with an update.

Also, thank you so much for the quick response, really appreciate the help!

**Reply by tmowbrey** [reply]
Rebooted both the MiniDSP and the WI-DG and have the same response: 

```
Sent: ReadHardwareId
Recv: HardwareId { payload: [01, 0c, 0a] }
Sent: ReadMemory { addr: ffa1, size: 01 }
```

**Reply by mrene** [reply]
Can you run `minidsp --tcp=192.168.1.183 -vvv` ?

Also, are you able to take a packet capture of both the plugin and minidsp-rs so I could compare the differences?

**Reply by tmowbrey** [reply]
Here's the response I get for that:

```
Sent: [31]
Sent: ReadHardwareId
Recv: [31, 01, 0c, 0a]
Recv: HardwareId { payload: [01, 0c, 0a] }
Sent: [05, ff, a1, 01]
Sent: ReadMemory { addr: ffa1, size: 01 }
```

Is there a method I can send my Wireshark data direct to you? I'd prefer not to post on a public forum as it contains general network traffic I'd rather not put here (I'm not adept at packet capture so didn't know how to limit what I was capturing).

**Reply by mrene** [reply]
Sure just email it to me mathieu.rene@gmail.com

**Reply by mrene** [reply]
I see what's going on - your wi-dg is using a different message framing strategy. I patched it to detect it in a hacky way, but it should work.

Can you try this version? If it works fine I'll make a new release.
Windows: https://github.com/mrene/minidsp-rs/suites/2575111061/artifacts/56143798 
Others: (at the very bottom of the page) https://github.com/mrene/minidsp-rs/actions/runs/781380099

**Reply by tmowbrey** [reply]
Doesn't look like it. I'm getting the following:

```
Sent: [31]
Sent: ReadHardwareId
thread 'tokio-runtime-worker' panicked at 'split_to out of bounds: 64 <= 0', C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\bytes-1.0.1\src\bytes_mut.rs:362:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Recv: [31, 01, 0c, 0a]
Recv: HardwareId { payload: [01, 0c, 0a] }
Sent: [05, ff, a1, 01]
Sent: ReadMemory { addr: ffa1, size: 01 }
```

**Reply by mrene** [reply]
Take two:
https://github.com/mrene/minidsp-rs/suites/2575180698/artifacts/56145523

**Reply by tmowbrey** [reply]
Success! You're awesome! Thank you :)

```
Sent: [31]
Sent: ReadHardwareId
Recv: [31, 01, 0c, 0a]
Recv: HardwareId { payload: [01, 0c, 0a] }
Sent: [05, ff, a1, 01]
Sent: ReadMemory { addr: ffa1, size: 01 }
Recv: [05, ff, a1, 64]
Recv: MemoryData(MemoryView { base: ffa1, data: [64] })
Sent: [05, ff, d8, 08]
Sent: ReadMemory { addr: ffd8, size: 08 }
Recv: [05, ff, d8, 00, 00, 00, 00, ff, ff, ff, ff]
Recv: MemoryData(MemoryView { base: ffd8, data: [00, 00, 00, 00, ff, ff, ff, ff] })
Sent: [14, 00, 44, 02]
Sent: ReadFloats { addr: 44, len: 02 } MasterStatus { preset: 0, source: Analog, volume: Gain(-0.0), mute: false }
Input levels: -74.8, -75.5
Output levels: -74.0, -120.0, -74.8, -120.0

(0x44 | 68) <> Meter02_C1_0
Recv: [14, 00, 44, cd, 87, 95, c2, 6b, e8, 96, c2]
Recv: FloatData(FloatView { base: 44, data: [-74.765236, -75.45394] })
Sent: [14, 00, 4a, 04]
Sent: ReadFloats { addr: 4a, len: 04 } (0x4a | 74) <> Meter10_C1_4
Recv: [14, 00, 4a, 50, 0a, 94, c2, 00, 00, f0, c2, 61, 93, 95, c2, 00, 00, f0, c2]
Recv: FloatData(FloatView { base: 4a, data: [-74.02014, -120.0, -74.78785, -120.0] })
```

**Comment by mrene** [comment]
Is this really a 2x4hd? The hardware id usually responds 0xa, and yours responds with 0xc. If it is, does it have the dirac upgrade?  

I think the dev version (not released yet) wouldn't recognize this as a 2x4hd, so I want to make sure it does :D

**Reply by tmowbrey** [reply]
Back of the device says "miniDSP 2x4HD" and invoice says "miniDSP 2x4HD -- Audio Processor in box, a USB DAC+DSP in a package powered by a powerful SHARC 400HMz floating point DSP". Purchased the device early this week and was just delivered from Hong Kong yesterday.

All to say, I'm pretty sure it's a 2x4HD :)

If there is something else I can check to verify something more details, happy to do that.

**Reply by mrene** [reply]
Thanks for double-checking! It might be a newer board version then :)
